bin/build.sh

```diff
@@ -1,73 +1,81 @@
 
-# Makes it so that all commands are displayed as they get executed
-set -x
-
-# WCA url and absolute path to WCA logo file
 wca_url="https://www.worldcubeassociation.org/"
 wca_docs_url="https://documents.worldcubeassociation.org/"
+# Absolute paths to the logo file, the main stylesheet and the edudoc header
 logo=$(realpath "./assets/WCAlogo_notext.svg")
-# Absolute paths to the stylesheets and the edudoc header
-stylesheet=$(realpath "./assets/style.css")
-edudoc_stylesheet=$(realpath "./assets/edudoc-style.css")
+main_stylesheet=$(realpath "./assets/style.css")
 edudoc_header=$(realpath "./assets/edudoc-header.html")
 # Current date
 compile_date=$(date '+%Y-%m-%d')
 
-# Function for converting Markdown files from a given folder into PDF files using the given stylesheet
-# The first argument ($1) is the folder and the second argument is the stylesheet ($2)
+# Function for converting Markdown files from a given directory into PDF files.
+# The first argument ($1) refers to the directory name (e.g: documents or edudoc).
 convert_to_pdf() {
+  rm -rf "build/$1"
   cp -r "$1/" build/
 
-  # Find formal and legal Markdown files and build PDFs out of them.
+  # Find Markdown files and build PDFs out of them.
   find "build/$1" -name '*.md' | while read file; do
-    echo "Processing $file using the ${1} stylesheet"
+    echo "Converting ${file#build/}..."
 
-    pdf_name="${file%.md}.pdf"
-    html_name="${file%.md}.html"
-    
-    file_headline=$(head -n 1 "$file")
-    document_title=$(echo "$file_headline" | sed -E "s/#+\s*//")
+    pdf_path="${file%.md}.pdf"
+    html_path="${file%.md}.html"
+    # Use first header as the document title; remove all # characters and whitespace from the beginning
+    document_title=$(head -n 1 "$file" | sed -E "s/^#+\s*//")
+    # Absolute path to the custom stylesheet
+    custom_stylesheet=$(realpath "./assets/$1-style.css")
 
     # Replace wca{...} and wcadoc{...} with absolute WCA URLs (the correct URLs only get inserted in production)
     sed -Ei "s#wca\{([^}]*)\}#$wca_url\1#g" "$file"
     sed -Ei "s#wcadoc\{([^}]*)\}#$wca_docs_url\1#g" "$file"
     # Replace {logo} with the path to the WCA logo in /assets
     sed -Ei "s#\{logo\}#$logo#g" "$file"
 
-    # This creates the custom header for each edudoc and saves it next to the doc that will use it
-    if [ $1 = "edudoc" ]; then
-      header_html="${file%.md}-header.html"
-    
+    # This creates the custom header for each edudoc and prepends it to each file
+    if [ "$1" = "edudoc" ]; then
       # Set the document title, path to the logo and the date
-      sed -E "s#DOCUMENT_TITLE#$document_title#g" "$edudoc_header" |
-      sed -E "s#WCA_LOGO_PATH#$logo#g" |
-      sed -E "s#DATE#$compile_date#g" > "$header_html"
+      sed -E "s#\{document_title\}#$document_title#g" "$edudoc_header" |
+      sed -E "s#\{wca_logo_path\}#$logo#g" |
+      sed -E "s#\{date\}#$compile_date#g" |
+      echo -e "$(cat /dev/stdin)\n\n$(cat $file)" > "$file"
```

Comment:
Am I correct in reading that you're basically pre-pending `stdin` to the existing file?

This `echo` is hard to read and understand, and manually typing out `\n` line breaks is never a good idea in scripts. If this actually does what I think it does, then consider some other approach along the lines of:
1. Piping the `sed` stuff into a new. temporary file using `echo > $temp_file`
2. Appending the original file using `cat $file >> $temp_file`
3. Renaming `$temp_file` to the original `$file` if necessary.

Reply:
That's exactly how it used to work. I felt like the temporary file was unnecessary, so I changed it. I don't think this part is that bad personally.

Reply:
Oh, but the new line characters are totally unnecessary, I'll get rid of them

Reply:
I still do feel quite strongly that `echo $(cat /dev/stdin)` is not a good idiom, especially when piping. Not sure what actual Bash gurus say about this, but unless you find a really compelling argument or an article that underlines that the way you're doing it is common best practice, I'd rather have it be separate echo statments like before.